workflow/util/util.go

```diff
@@ -331,13 +332,36 @@ func ResumeWorkflow(wfIf v1alpha1.WorkflowInterface, workflowName string) error
 			wf.Spec.Suspend = nil
 			updated = true
 		}
-		// To resume a workflow with a suspended node we simply mark the node as Successful
+
 		for nodeID, node := range wf.Status.Nodes {
 			if node.Type == wfv1.NodeTypeSuspend && node.Phase == wfv1.NodeRunning {
-				node.Phase = wfv1.NodeSucceeded
-				node.FinishedAt = metav1.Time{Time: time.Now().UTC()}
-				wf.Status.Nodes[nodeID] = node
-				updated = true
+				//if we have a selector then we only unsuspend matching nodes
+				stepMatchesSelector := false
+				if stepSelector == nil || stepSelector.Empty() {
+					stepMatchesSelector = true
+				} else {
+					inputsMap := fields.Set{}
+					for _, param := range node.Inputs.Parameters {
+						if param.Value != nil {
+							inputsMap[param.Name] = *param.Value
+						}
+					}
+
+					stepMatchesSelector = stepSelector.Matches(inputsMap)
+				}
```

Comment:
I'm not too much of a fan of using the input parameters as arguments for selecting which step to resume; although I admit there is no obvious alternative.

For your use case -- would the entity calling `resume` have access to the index of the sub step? E.g., if you'd want to fail/resume `baz` in the following example would the entity know to call something like `argo resume loops-dag-9mdw9 --step B.2`?
```
Name:                loops-dag-9mdw9
Namespace:           argo
ServiceAccount:      default
Status:              Running (Suspended)
Created:             Mon Jan 06 10:20:38 -0800 (10 seconds ago)
Started:             Mon Jan 06 10:20:38 -0800 (10 seconds ago)
Duration:            10 seconds

STEP                            PODNAME  DURATION  MESSAGE
 ● loops-dag-9mdw9 (loops-dag)
 ├-● B(0:foo) (whalesay)
 ├-● B(1:bar) (whalesay)
 └-● B(2:baz) (whalesay)
```

If not maybe we could allow for both parameter-based selectors and step indices.

Reply:
It wouldn't generally know the index. A typical use of this could be having two steps, the first being a pod that kicks off a background process and the second being the suspend step. 

The suspend step would have an input parameter that comes from the pod that kicks off the background process, so when the background process is complete and the suspend step is resumed, it can request the resume without introspecting the workflow.

Is referring to steps by index used anywhere else in the codebase? It seems that logic could be quite complex ...

Reply:
Hmmm, in that case maybe my suggestion would be something like this. We could create a new field under suspend that would allow users to "tag" a suspend step. Maybe something like:
```yaml
  - name: whalesay
    inputs:
      parameters:
      - name: message
    suspend:
      tag: "{{inputs.parameters.message}}"
```

Then the step could be resumed with: `argo resume --step-tag [TAG]`

This could provide the benefit of having a consistent, user-defined tag to unsuspend the workflow from an external, automated process that don't have access to the workflow. What do you think?

cc @alexec

Reply:
Yes, that would work as an alternative. It does feel to me like it adds quite a bit more code complexity, but I understand the benefits so I'll implement it if that's the direction you want to go.

Reply:
Let's wait and see what Alex/Jesse have to say about this before you start the implementation.

Reply:
(@dmayle Just FYI, we're a bit busy with the upcoming API Server release and will get to this a couple of days after the release)

Reply:
what about something similar to Kubernetes `field-selector`, I'd call this `nodeSelector`:

```
func (n Nodes) FindByDisplayName(name string) *NodeStatus {
	return n.FindBySelector(fields.ParseSelectorOrDie("displayName=" + name))
}

func (n Nodes) FindBySelector(s fields.Selector) *NodeStatus {
	for _, i := range n {
		f := fields.Set{
			"displayName": i.DisplayName,
		}
		if s.Matches(f) {
			return &i
		}
	}
	return nil
}
```